### PR TITLE
AVRO-2413: Refactor io.Validate to use a mapping

### PR DESCRIPTION
**Pull request by kojiromike**
Here is a refactor of io.Validate to reduce its complexity and make it more readable.

### CODE REVIEW

**Comment by sireliah** [comment]
I really like your initiative to simplify the validation @kojiromike . However I still would like to keep the extra information about what exactly was wrong with the datum, as defined here: https://github.com/apache/avro/pull/306

Do you think it's possible to keep both changes?

**Comment by kojiromike** [comment]
@sireliah Definitely! I mean, let's get one of them merged and then work on the other, what do you say?

The way I see it, your changeset could probably be re-implemented as some kind of cross-cutting decorator (literally) of the validation methods that I'm implementing as lambdas here.

File: lang/py3/avro/io.py [hunk 1]
```diff
@@ -102,51 +123,9 @@ def Validate(expected_schema, datum):
...
-            return False
-    for datum_field in datum.keys():
-        if datum_field not in expected_schema_field_names:
-            return False
-    return True
-  else:
+  try:
+    return _valid[expected_schema.type](expected_schema, datum)
+  except KeyError:
     raise AvroTypeException('Unknown Avro schema type: %r' % schema_type)
```
Comment (sireliah) [inline lang/py3/avro/io.py]: Not related to your changes here, but `AvroTypeException` will fail here, because it expects 3 arguments, and gets only one.

**Comment by sireliah** [comment]
This is very good idea @kojiromike! (I'm not even able to squash the commits after so many merges at this point :D)  Let's begin with your PR then.

As for the improved exceptions we'll need two components.
1. Aggregate errors in some data structure when going through all calls of the `Validate()` func. The decorator could be class based to collect somehow exceptions from all recursive calls. I don't know yet how to do it, I'll think about this.
2. We need to place human readable texts somewhere. Maybe the decorator could own those definitions? For sure, we should be able to explain *what* was validated negatively.

P.S. I see that in your PR what would work best is the pattern matching. Not in Python (yet) unfortunately. : ]

**Comment by kojiromike** [comment]
@sireliah it’s true that python doesn’t have true pattern matching, but have you looked at single dispatch generic functions? Can we make those work?

**Comment by sireliah** [comment]
I never had opportunity, but this sounds promising. I'll experiment with this.

**Comment by kojiromike** [comment]
@Fokko this is passing, too, if you have time to take a look.

File: lang/py3/avro/io.py (hunk 1 above)
Comment (Fokko) [reply lang/py3/avro/io.py]: Good catch @sireliah, can you make a ticket for this issue?

Comment (sireliah) [reply lang/py3/avro/io.py]: Here! https://issues.apache.org/jira/browse/AVRO-2443

**Comment by Fokko** [comment]
Conflicts 😭

**Pull request merged**